# **NYC TLC Basic Data Exploration Project**

The New York City Taxi and Limousine Commission (New York City TLC) has shared their data.

To get clear insights, New York TLC's data must be analyzed, key variables identified, and the dataset ensured it is ready for analysis.

# Course 2 End-of-course project: Inspect and analyze data

In this activity, you will examine data provided and prepare it for analysis.  
<br/>   

**The purpose** of this project is to investigate and understand the data provided.
  
**The goal** is to use a dataframe constructed within Python, perform a cursory inspection of the provided dataset, and inform team members of your findings. 


* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities.

* Compile summary information about the data to inform next steps.

* Use insights from your examination of the summary data to guide deeper investigation into specific variables.


# **Identify data types and relevant variables using Python**


In [1]:
#import a few libraries
import pandas as pd
import numpy as np
# Import data frame
df = pd.read_csv('2017_Yellow_Taxi_Trip_Data.csv')

In [2]:
#==> ENTER YOUR CODE HERE
df.head(10)

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0    24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1    35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2   106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3    38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4    30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   
5    23345809         2   03/25/2017 8:34:11 PM   03/25/2017 8:42:11 PM   
6    37660487         2   05/03/2017 7:04:09 PM   05/03/2017 8:03:47 PM   
7    69059411         2   08/15/2017 5:41:06 PM   08/15/2017 6:03:05 PM   
8     8433159         2   02/04/2017 4:17:07 PM   02/04/2017 4:29:14 PM   
9    95294817         1   11/10/2017 3:20:29 PM   11/10/2017 3:40:55 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
4                1           4.37           1                  N   
5                6           2.30           1                  N   
6                1          12.83           1                  N   
7                1           2.98           1                  N   
8                1           1.20           1                  N   
9                1           1.60           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
4             4           112             2         16.5    0.5      0.5   
5           161           236             1          9.0    0.5      0.5   
6            79           241             1         47.5    1.0      0.5   
7           237           114             1         16.0    1.0      0.5   
8           234           249             2          9.0    0.0      0.5   
9           239           237             1         13.0    0.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  
0        2.76           0.0                    0.3         16.56  
1        4.00           0.0                    0.3         20.80  
2        1.45           0.0                    0.3          8.75  
3        6.39           0.0                    0.3         27.69  
4        0.00           0.0                    0.3         17.80  
5        2.06           0.0                    0.3         12.36  
6        9.86           0.0                    0.3         59.16  
7        1.78           0.0                    0.3         19.58  
8        0.00           0.0                    0.3          9.80  
9        2.75           0.0                    0.3         16.55

In [3]:
#==> ENTER YOUR CODE HERE
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

Unusually clean dataset. all non-null values. 
<br>tpep_pickup_datetime and dropoff_datetime need to be changed to date time objects. <br>
See if creating a trip duration will be needed in any future project tasks.<br> 
all other data types look appropriate

In [14]:
#convert date time objects
df['tpep_pickup_datetime'] = pd.to_datetime(df['tpep_pickup_datetime'])
df['tpep_dropoff_datetime'] = pd.to_datetime(df['tpep_dropoff_datetime'])
#check dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Unnamed: 0             22699 non-null  int64         
 1   VendorID               22699 non-null  int64         
 2   tpep_pickup_datetime   22699 non-null  datetime64[ns]
 3   tpep_dropoff_datetime  22699 non-null  datetime64[ns]
 4   passenger_count        22699 non-null  int64         
 5   trip_distance          22699 non-null  float64       
 6   RatecodeID             22699 non-null  int64         
 7   store_and_fwd_flag     22699 non-null  object        
 8   PULocationID           22699 non-null  int64         
 9   DOLocationID           22699 non-null  int64         
 10  payment_type           22699 non-null  int64         
 11  fare_amount            22699 non-null  float64       
 12  extra                  22699 non-null  float64       
 13  m

In [13]:
#==> ENTER YOUR CODE HERE
df.describe()

Unnamed: 0      VendorID  passenger_count  trip_distance  \
count  2.269900e+04  22699.000000     22699.000000   22699.000000   
mean   5.675849e+07      1.556236         1.642319       2.913313   
std    3.274493e+07      0.496838         1.285231       3.653171   
min    1.212700e+04      1.000000         0.000000       0.000000   
25%    2.852056e+07      1.000000         1.000000       0.990000   
50%    5.673150e+07      2.000000         1.000000       1.610000   
75%    8.537452e+07      2.000000         2.000000       3.060000   
max    1.134863e+08      2.000000         6.000000      33.960000   

         RatecodeID  PULocationID  DOLocationID  payment_type   fare_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
mean       1.043394    162.412353    161.527997      1.336887     13.026629   
std        0.708391     66.633373     70.139691      0.496211     13.243791   
min        1.000000      1.000000      1.000000      1.000000   -120.000000   
25%        1.000000    114.000000    112.000000      1.000000      6.500000   
50%        1.000000    162.000000    162.000000      1.000000      9.500000   
75%        1.000000    233.000000    233.000000      2.000000     14.500000   
max       99.000000    265.000000    265.000000      4.000000    999.990000   

              extra       mta_tax    tip_amount  tolls_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000   
mean       0.333275      0.497445      1.835781      0.312542   
std        0.463097      0.039465      2.800626      1.399212   
min       -1.000000     -0.500000      0.000000      0.000000   
25%        0.000000      0.500000      0.000000      0.000000   
50%        0.000000      0.500000      1.350000      0.000000   
75%        0.500000      0.500000      2.450000      0.000000   
max        4.500000      0.500000    200.000000     19.100000   

       improvement_surcharge  total_amount  
count           22699.000000  22699.000000  
mean                0.299551     16.310502  
std                 0.015673     16.097295  
min                -0.300000   -120.300000  
25%                 0.300000      8.750000  
50%                 0.300000     11.800000  
75%                 0.300000     17.800000  
max                 0.300000   1200.290000

We have 2 vendors
<br>Pickup and dropoff columns were omitted from describe()
<br>passenger counts ranging from 1-6
<br>trip distance ranging from 0 to 33.95 miles **investigate further**
<br>Locations are not geolocations, just **encoded labels**
<br>payment type ranges from 1 to 4, **label encoding**
<br>fare amount ranges from -120 to 999.99. **investigate further**
<br>extra ranges from -1 to 4.5
<br>mta tax is uniform
<br>tip amount ranges from 0 to 200
<br>tolls vary from 0 to 19.1
<br>improvement surcharge is uniform
<br>total_amount ranges from -120.3 to 1200.29 **investigate further**

### **Understand the data - Investigate the variables**


In [2]:
# Sort the data by trip distance descending
trip_distance_sorted_desc = df.sort_values(by = ['trip_distance'], ascending = False)
trip_distance_sorted_desc.head(20)

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
9280     51810714         2  06/18/2017 11:33:25 PM  06/19/2017 12:12:38 AM   
13861    40523668         2   05/19/2017 8:20:21 AM   05/19/2017 9:20:30 AM   
6064     49894023         2  06/13/2017 12:30:22 PM   06/13/2017 1:37:51 PM   
10291    76319330         2  09/11/2017 11:41:04 AM  09/11/2017 12:18:58 PM   
29       94052446         2   11/06/2017 8:30:50 PM  11/07/2017 12:00:00 AM   
18130    90375786         1   10/26/2017 2:45:01 PM   10/26/2017 4:12:49 PM   
5792     68023798         2   08/11/2017 2:14:01 PM   08/11/2017 3:17:31 PM   
15350    77309977         2   09/14/2017 1:44:44 PM   09/14/2017 2:34:29 PM   
10302    43431843         1   05/15/2017 8:11:34 AM   05/15/2017 9:03:16 AM   
2592     51094874         2   06/16/2017 6:51:20 PM   06/16/2017 7:41:42 PM   
20612    67238347         2   08/08/2017 9:01:00 PM   08/08/2017 9:40:04 PM   
1908     81396251         2  09/27/2017 11:49:52 PM  09/28/2017 12:26:29 AM   
20545    71466504         1   08/24/2017 1:03:49 PM   08/24/2017 1:49:05 PM   
4138    100097208         2   11/26/2017 3:37:35 PM   11/26/2017 4:50:22 PM   
15169    71224174         2   08/23/2017 4:20:15 PM   08/23/2017 5:08:55 PM   
1496     27465882         2   04/03/2017 3:37:53 PM   04/03/2017 4:44:33 PM   
7217     13572254         1   02/20/2017 2:28:11 PM   02/20/2017 3:29:14 PM   
908      25075013         2   03/27/2017 1:01:38 PM   03/27/2017 1:38:44 PM   
19483    34486843         2   04/26/2017 8:31:31 AM  04/26/2017 11:10:50 AM   
4715     78418799         2   09/17/2017 8:04:24 PM   09/17/2017 9:02:24 PM   

       passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
9280                 2          33.96           5                  N   
13861                1          33.92           5                  N   
6064                 1          32.72           3                  N   
10291                1          31.95           4                  N   
29                   1          30.83           1                  N   
18130                1          30.50           1                  N   
5792                 1          30.33           2                  N   
15350                1          28.23           2                  N   
10302                1          28.20           2                  N   
2592                 1          27.97           2                  N   
20612                1          27.88           1                  N   
1908                 2          27.34           1                  N   
20545                1          27.20           1                  N   
4138                 2          26.86           1                  N   
15169                2          26.54           1                  N   
1496                 1          26.39           2                  N   
7217                 1          26.20           1                  N   
908                  2          26.12           4                  N   
19483                1          26.12           2                  N   
4715                 1          25.86           4                  N   

       PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
9280            132           265             2       150.00    0.0      0.0   
13861           229           265             1       200.01    0.0      0.5   
6064            138             1             1       107.00    0.0      0.0   
10291           138           265             2       131.00    0.0      0.5   
29              132            23             1        80.00    0.5      0.5   
18130           132           220             1        90.50    0.0      0.5   
5792            132           158             1        52.00    0.0      0.5   
15350            13           132             1        52.00    0.0      0.5   
10302            90           132             1        52.00    0.0      0.5   
2592            261           132 

In [3]:
#sort trip distance ascending
trip_distance_sorted_asc = df.sort_values(by = ['trip_distance'], ascending = True)
trip_distance_sorted_asc.head(20)

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
22026    63642923         1   07/27/2017 7:44:24 AM   07/27/2017 7:44:24 AM   
795     101135030         1   11/30/2017 7:11:34 AM   11/30/2017 7:11:34 AM   
6908     24162045         2   03/26/2017 2:07:08 AM   03/26/2017 2:07:12 AM   
13561    14504365         1   02/23/2017 4:06:31 PM   02/23/2017 4:06:54 PM   
12238    95544923         1   11/11/2017 9:28:13 AM   11/11/2017 9:28:27 AM   
6878     69466211         2   08/17/2017 4:06:26 AM   08/17/2017 4:06:29 AM   
14372   108639975         1   12/22/2017 1:39:55 PM   12/22/2017 1:44:23 PM   
20638   106367018         1   12/15/2017 4:09:43 PM   12/15/2017 4:09:43 PM   
920      92102517         2   10/31/2017 8:29:53 PM   10/31/2017 8:30:27 PM   
16263     5716905         1   01/27/2017 8:58:29 PM   01/27/2017 8:58:50 PM   
19636     8288479         2   02/04/2017 6:44:24 AM   02/04/2017 6:44:25 AM   
15295    83531047         1  10/04/2017 10:07:50 PM  10/04/2017 10:07:50 PM   
12144   109947527         2   12/28/2017 8:13:08 AM   12/28/2017 8:13:35 AM   
10281    55302347         2   06/05/2017 5:34:25 PM   06/05/2017 5:36:29 PM   
21130    82975077         2   10/03/2017 8:03:35 AM   10/03/2017 8:03:48 AM   
20174    61310769         1   07/19/2017 9:39:40 AM   07/19/2017 9:47:10 AM   
22190    87078850         1   10/16/2017 8:34:07 AM   10/16/2017 8:34:10 AM   
21141    71819090         1   08/25/2017 6:51:00 PM   08/25/2017 6:51:59 PM   
1449     91160051         2   10/28/2017 7:41:58 PM   10/28/2017 7:42:11 PM   
14165    91282423         2   10/29/2017 1:14:35 AM   10/29/2017 1:18:13 AM   

       passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
22026                1            0.0           1                  N   
795                  1            0.0           1                  N   
6908                 1            0.0           5                  N   
13561                2            0.0           5                  N   
12238                2            0.0           1                  N   
6878                 4            0.0           5                  N   
14372                1            0.0           1                  N   
20638                1            0.0           1                  N   
920                  1            0.0           1                  N   
16263                1            0.0           1                  N   
19636                2            0.0           5                  N   
15295                1            0.0           1                  N   
12144                1            0.0           1                  N   
10281                2            0.0           1                  N   
21130                1            0.0           1                  N   
20174                1            0.0           1                  N   
22190                1            0.0           1                  N   
21141                1            0.0           1                  N   
1449                 1            0.0           2                  N   
14165                1            0.0           5                  N   

       PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
22026            41           264             2         10.5    0.0      0.5   
795             246           264             2          8.0    0.0      0.5   
6908             61            61             1         18.0    0.0      0.0   
13561           175           175             3         32.0    0.0      0.0   
12238           145           145             2          2.5    0.0      0.5   
6878            100           100             1         98.5    0.0      0.5   
14372           226           226             2          4.5    0.0      0.5   
20638            43           264             2          2.5    1.0      0.5   
920              50            50             2          2.5    0.5      0.5   
16263           249           249 

The total amounts are all over the place for trips with 0 recorded distance.

In [50]:
#create a new data frame that filters for trips with 0 distance, and high total fares
mask_trip_distance_zero_high_total_amount = (df['trip_distance'] == 0.00) & (df['total_amount'] > 100.00)

df[mask_trip_distance_zero_high_total_amount][['trip_distance', 'total_amount']].sort_values(by = ('total_amount'), ascending = False)


trip_distance  total_amount
20312            0.0        450.30
12511            0.0        233.74
15474            0.0        211.80
11269            0.0        151.82
7281             0.0        120.96
13359            0.0        111.95

These may be errors in the data. We will want to explore this further in EDA, and contact the shareholders regarding these data points.

In [18]:
# Sort the data by total amount and print the top 20 values
total_amount_sorted = df.sort_values(by = ['total_amount'], ascending = False)

total_amount_sorted.head(20)


8476     1200.29
20312     450.30
13861     258.21
12511     233.74
15474     211.80
6064      179.06
16379     157.06
3582      152.30
11269     151.82
9280      150.30
1928      137.80
10291     131.80
6708      126.00
11608     123.30
908       121.56
7281      120.96
18130     119.31
13621     115.94
13359     111.95
29        111.38
Name: total_amount, dtype: float64

Note outliers to pay attention to

In [19]:
total_amount_sorted_asc = df['total_amount'].sort_values(ascending = True)
# Sort the data by total amount and print the bottom 20 values
total_amount_sorted_asc.head(20)

12944   -120.30
20698     -5.80
17602     -5.80
11204     -5.30
14714     -5.30
8204      -4.80
20317     -4.80
10281     -4.30
5448      -4.30
4423      -4.30
18565     -3.80
314       -3.80
5758      -3.80
1646      -3.30
10506      0.00
4402       0.00
5722       0.00
22566      0.00
19067      0.30
14283      0.31
Name: total_amount, dtype: float64

Noting negative fare amounts to keep our eye on during EDA, and potentially communicate to stakeholders.

payment types:
    1: credit card
    2: cash
    3: no charge
    4: dispute

In [18]:
#check if disputed or no charge payments are the cause of these negative values
dispute_df = df[(df['payment_type'] == 4) | (df['payment_type'] == 3)]
#sort by total amount ascending
dispute_df['total_amount'].sort_values(ascending = True).head(20)

12944   -120.30
17602     -5.80
20698     -5.80
14714     -5.30
11204     -5.30
8204      -4.80
20317     -4.80
10281     -4.30
5448      -4.30
4423      -4.30
18565     -3.80
5758      -3.80
314       -3.80
1646      -3.30
19067      0.30
14283      0.31
5207       3.30
17747      3.30
22190      3.30
17579      3.30
Name: total_amount, dtype: float64

Filtering for disputed and no charge payment type we can see where all these negative values are coming from. 

In [21]:
#explore payment type column
df['payment_type'].value_counts()

1    15265
2     7267
3      121
4       46
Name: payment_type, dtype: int64

In [ ]:
as a reminder:
<br>payment types:
    1: credit card
    2: cash
    3: no charge
    4: dispute
<br> credit cards and cash make up the vast majority of payment types.

Investigate tips for different payments types

In [35]:
#create a boolean mask to pull out cash and credit card types
mask_pt_1 = df['payment_type'] == 1
#average tip amount for credit cards
print(df[mask_pt_1]['tip_amount'].mean())


# average tip for cash
mask_pt_2 = df['payment_type'] == 2
print(df[mask_pt_2]['tip_amount'].mean())

2.7298001965279934
0.0


Cash tips are not reported

Same thing, but computed with the .mean() method

In [51]:
# What is the average tip for trips paid for with credit card?
avg_cc_tip = df[df['payment_type']==1]['tip_amount'].mean()
print('Avg. cc tip:', avg_cc_tip)

# What is the average tip for trips paid for with cash?
avg_cash_tip = df[df['payment_type']==2]['tip_amount'].mean()
print('Avg. cash tip:', avg_cash_tip)

Avg. cc tip: 2.7298001965279934
Avg. cash tip: 0.0


Are there any tips other than credit card payments?

In [40]:
df.groupby('payment_type').mean()['tip_amount']

payment_type
1    2.7298
2    0.0000
3    0.0000
4    0.0000
Name: tip_amount, dtype: float64

In [22]:
#What are the vendorID counts?
df['VendorID'].value_counts()


2    12626
1    10073
Name: VendorID, dtype: int64

In [36]:
# do one of the vendors make more money?
df.groupby('VendorID').mean()['total_amount']

VendorID
1    16.298119
2    16.320382
Name: total_amount, dtype: float64

In [52]:
df.groupby(['VendorID']).mean(numeric_only=True)[['total_amount']]

total_amount
VendorID              
1            16.298119
2            16.320382

In [53]:
#count credit card payments among all passenger counts
credit_card = df[df['payment_type']==1]
credit_card['passenger_count'].value_counts()

1    10977
2     2168
5      775
3      600
6      451
4      267
0       27
Name: passenger_count, dtype: int64

In [39]:
# Calculate the average tip amount for each passenger count (credit card payments only)
mask_cc_only = df['payment_type'] == 1
df[mask_cc_only].groupby('passenger_count').mean()['tip_amount']

passenger_count
0    2.610370
1    2.714681
2    2.829949
3    2.726800
4    2.607753
5    2.762645
6    2.643326
Name: tip_amount, dtype: float64

In [54]:
credit_card.groupby(['passenger_count']).mean(numeric_only=True)[['tip_amount']]

tip_amount
passenger_count            
0                  2.610370
1                  2.714681
2                  2.829949
3                  2.726800
4                  2.607753
5                  2.762645
6                  2.643326

## Preliminary insights derived from the data
The data is unusually clean. No null values.
<br> I only had to convert 2 date time columns into datetime objects
<br> We may need to create a time delta column if we have requests regarding trip duration.


#### Unusual or questionable values
There are trip distance values of 0.00. explore this futher in EDA
<br>There are total amount values of 0.00 and negative values for disputed or no charge payment type. There are a small number of negative values in this large dataset, and this is part of running a taxi business. We probably want to keep this data. Consult with the team on this.

#### How many of each payment type are represented in the data
1    15265
2     7267
3      121
4       46
Name: payment_type, dtype: int64

#### Mean tip amount for each payment type
payment_type
1    2.7298
2    0.0000
3    0.0000
4    0.0000
Name: tip_amount, dtype: float64

#### How many rides each vendor provided
2    12626
1    10073
Name: VendorID, dtype: int64

#### Mean total amount for each vendor
VendorID
1    16.298119
2    16.320382
Name: total_amount, dtype: float64

#### Mean tip amount for each passenger count (only of those who paid by credit card)
passenger_count
0    2.610370
1    2.714681
2    2.829949
3    2.726800
4    2.607753
5    2.762645
6    2.643326
Name: tip_amount, dtype: float64
